subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

```diff
@@ -89,6 +89,63 @@ See the release notes for link:https://github.com/JetBrains/kotlin/releases/tag/
 
 === Deprecations
 
+==== Deprecated `Project.buildDir` is to be replaced by `Project.layout.buildDirectory`
+
+The `Project.buildDir` property is deprecated.
+It uses eager APIs and has ordering issues if the value is read in build logic and then later modified as outputs could end up in different locations.
+
+It is replaced by a `link:{javadocPath}/org/gradle/api/file/DirectoryProperty.html[DirectoryProperty]` found at `Project.layout.buildDirectory`.
+See the `link:{dslPath}/org.gradle.api.file.ProjectLayout.html[ProjectLayout]` interface for details.
+
+Note that at this stage, Gradle will not print deprecation warnings if you still use `Project.buildDir`.
+We know this is a big change and want to give time for authors of major plugin to move away from its usage first.
+
+The switch from a `File` to a `DirectoryProperty` requires adaptations in build logic.
+The main impact is that you cannot use the property inside a `String` to expand it.
+Instead, you should leverage the `dir` and `file` methods to compute the location you want.
+
+Code like the following, either in a build script or a task configuration block:
```

Comment:
🤔  I think it would be good to show how to migrate when what is expected is a `FilesystemLocation` or a `File` or a `String` path. Explaining this explicitly will reduce the amount of support we will have to do on this change.

Reply:
+1